network/p2p/topic_validator.go

```diff
@@ -0,0 +1,26 @@
+package p2p
+
+import (
+	"context"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+
+	"github.com/onflow/flow-go/model/flow"
+)
+
+type StakedValidator struct {
+	getIdentity func(peer.ID) (*flow.Identity, bool)
+}
+
+func (v *StakedValidator) Validate(ctx context.Context, receivedFrom peer.ID, msg *pubsub.Message) pubsub.ValidationResult {
+	// check that message contains a valid sender ID
+	if from, err := messageSigningID(msg); err == nil {
```

Comment:
1. Since `messageSigningID` is no longer exclusive to `readSubscription` can we move it to `libp2pUtils`?
2. Minor optimization - `messageSigningID` call will repeated in `readSubscription`. I noticed that libp2p pubsub message offers this field - [ValidatorData](https://github.com/libp2p/go-libp2p-pubsub/blob/master/pubsub.go#L215). We can stash the `peerID` in there and check for its presence in `readSubscription`.
3. How about a refactor-> rename to `signerIDFromMessage` ?

Reply:
on 2.: the messageSigningID is really supposed to be a cheap operation: look up a couple of fields, essentially. That's despite the scary name of `UnmarshalPublicKey` (which isn't actually costly).
 
I think there's a risk to over-optimize this.

Reply:
@vishalchangrani this is great to know! very useful feature tbh, could have been used to implement origin validator as a topic validator instead of in `readSubscription`.

Reply:
However, I think atm this is unnecessary, because I am hoping that https://github.com/onflow/flow-go/pull/1167 will eventually get adopted in some form, which means we will no longer need the origin validation.